ipa-core/src/protocol/ipa_prf/quicksort.rs

```diff
@@ -54,65 +60,93 @@ where
-                })),
-            )
-            .try_collect()
-            .await?;
+    let mut ranges_for_next_pass = Vec::with_capacity(ranges_to_sort.len() * 2);
+    let mut quicksort_pass = 1;
+
+    // iterate through all of the potentially incorrectly ordered ranges
+    // make one pass, comparing each element to the pivot and splitting into two more
+    // potentially incorrectly ordered ranges
+    while !ranges_to_sort.is_empty() {
+        // compute the total number of comparisons that will be needed
+        let mut num_comparisons_needed = 0;
+        for range in &ranges_to_sort {
+            if range.len() > 1 {
+                num_comparisons_needed += range.len() - 1;
+            }
+        }
+
+        let c = ctx
+            .narrow(&Step::QuicksortPass(quicksort_pass))
+            .set_total_records(num_comparisons_needed);
+        let cmp_ctx = c.narrow(&Step::Compare);
+        let rvl_ctx = c.narrow(&Step::Reveal);
+
+        let comp: BitVec<usize, Lsb0> = seq_join(
+            ctx.active_work(),
+            stream_iter(
+                ranges_to_sort
+                    .iter()
+                    .filter(|r| r.len() > 1)
+                    .flat_map(|range| {
+                        // set up iterator
+                        let mut iterator = list[range.clone()].iter().map(get_key);
+                        // first element is pivot, apply key extraction function f
+                        let pivot = iterator.next().unwrap();
+                        zip(repeat(pivot), iterator)
+                    })
+                    .enumerate()
+                    .map(|(i, (pivot, k))| {
+                        let cmp_ctx = cmp_ctx.clone();
+                        let rvl_ctx = rvl_ctx.clone();
+                        let record_id = RecordId::from(i);
+                        async move {
+                            // Compare the current element against pivot and reveal the result.
+                            let comparison = compare_gt(cmp_ctx, record_id, k, pivot)
+                                .await?
+                                .reveal(rvl_ctx, record_id) // reveal outcome of comparison
+                                .await?;
+
+                            // desc = true will flip the order of the sort
+                            Ok::<_, Error>(Boolean::from(desc) == comparison)
+                        }
+                    }),
+            ),
+        )
+        .try_collect()
+        .await?;
+
+        let mut n = 0;
+        for range in &ranges_to_sort {
+            if range.len() <= 1 {
+                continue;
+            }
 
             // swap elements based on comparisons
             // i is index of first element larger than pivot
-            let mut i = b_l + 1;
-            for (j, b) in comp.iter().enumerate() {
+            let mut i = range.start + 1;
+            for (j, b) in comp[n..(n + range.len() - 1)].iter().enumerate() {
```

Comment:
I'm not clear on why you are iterating this way rather than having an iterator for the entirety of `comp` and doing: `for i in range.iter().skip(1) { if comp_iter.next() == Some(true) { swap } }` or similar.

Reply:
That's a great idea! I think i can make that work! I'll give it a try.

Reply:
What makes this very tricky is that the ranges are not necessarily adjacent, or tightly packed. Also, the indices of "comp" *are* tightly packed going from 0 upwards (they needed to be record_ids...).

Reply:
Hmm, so maybe you can iterate comp all at once, then co-iterate the ranges to find the indices you need.

```rust
let mut ranges = ranges.iter();
let mut range_it = ranges.next().unwrap(); // has to be at least one
let mut i = range_it.next().unwrap() + 1;  // first element of the range
for comparison in comp {
   let index = if let Some(n) = range_it.next() { n } else {
      range_it = ranges.next().unwrap().iter(); // has to be more here too
      // some other stuff to setup `i`
   };
   if comparison { swap }
}
```

Reply:
I'm attempting to do this here: https://github.com/private-attribution/ipa/pull/895/files

Still not working yet, but do you like the way this code looks better? I think it's slightly longer now.